### ISSUE TITLE
TC search/unification sometimes fails to take universe cumulativity into account

### ISSUE THREAD

**Issue by Janno** [coq/coq#14734]
#### Description of the problem

The two modules, differing only in the universe constraints introduced by `Definition Y`, show different behavior during TC search in the `Goal ..` proof script(s). To make matters more confusing, the difference disappears if we start the TC search without introducing the variable `A`.

```coq
#[universes(cumulative=yes)]
 Polymorphic Class t@{+s} (S : Type@{s}) (A B : Prop) :=
  T {
      prf : S -> A -> B;
    }.
Monomorphic Inductive X@{s | Set < s} : Type@{s} := x.

#[refine]
 Instance t_X_refl {A} : t X A A := {|prf := _|}.
Proof. auto. Qed.

Module A.
  (* Note universe constraint! *)
  Monomorphic Definition Y@{s | X.s < s} : Type@{s} := X.

  Goal forall A, t Y A A.
    Fail intros; typeclasses eauto. (* fails but should succeed, I think *)
    typeclasses eauto.    (* unexpectedly succeeds *)
  Qed.
End A.


Module B.
  (* No universe constraint this time! *)
  Monomorphic Definition Y@{s} : Type@{s} := X.

  Goal forall A, t Y A A.
    Fail Fail intros; typeclasses eauto. (* succeeds now *)
    typeclasses eauto.                   (* succeeds as before *)
  Qed.
End B.
```

#### Coq Version

8.13.2

**Comment by SkySkimmer** [comment]
No need to use a dummy Y, `Lemma foo@{s|X.s < s} : forall A, t@{s} X A A.` has the first behaviour and without the constraint it has the second behaviour.

**Comment by SkySkimmer** [comment]
Looks like a duplicate of https://github.com/coq/coq/issues/6976

**Comment by SkySkimmer** [comment]
It works before intro because unification is asked `(forall A, t@{X.s} X A A) == (forall A, t@{s} X A A)` (with differing univs) and uses conversion for it.
After intro we have to do actual unification `t@{X.s} X ? ? == t@{s} X A A` and conversion is called on the head subterms (`t` with differing universes).

Maybe worth reviving https://github.com/coq/coq/pull/12624 ?

**Comment by Janno** [comment]
If it fixes this issue and #12624 and also #13980 and probably some others then it would certainly be worth it. I fear the new unification engine is not going to land anytime soon. And I don't think there is way to tell TC search to use evarconv, or is there?
